components/public-api-server/pkg/idp/idp.go

```diff
@@ -0,0 +1,167 @@
+// Copyright (c) 2023 Gitpod GmbH. All rights reserved.
+// Licensed under the GNU Affero General Public License (AGPL).
+// See License.AGPL.txt in the project root for license information.
+
+package idp
+
+import (
+	"context"
+	"crypto/rand"
+	"crypto/rsa"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/http/httputil"
+	"net/url"
+	"time"
+
+	"github.com/gitpod-io/gitpod/common-go/log"
+	"github.com/go-chi/chi/v5"
+	"github.com/zitadel/oidc/pkg/crypto"
+	"github.com/zitadel/oidc/pkg/oidc"
+	"github.com/zitadel/oidc/pkg/op"
+	"gopkg.in/square/go-jose.v2"
+)
+
+func NewService(issuerBaseURL string, keyCache KeyCache) (*Service, error) {
+	idpKey, err := rsa.GenerateKey(rand.Reader, 2048)
+	if err != nil {
+		return nil, fmt.Errorf("cannot produce IDP private key: %w", err)
+	}
+	err = keyCache.Set(context.Background(), idpKey)
+	if err != nil {
+		return nil, fmt.Errorf("cannot cache IDP key: %w", err)
+	}
+
+	tokenEncryptionCode := make([]byte, 128)
+	_, err = io.ReadFull(rand.Reader, tokenEncryptionCode)
+	if err != nil {
+		return nil, fmt.Errorf("cannot produce random token encryption code: %w", err)
+	}
+
+	return &Service{
+		IssuerBaseURL:       issuerBaseURL,
+		TokenEncryptionCode: tokenEncryptionCode,
+		keys:                keyCache,
+	}, nil
+}
+
+type Service struct {
+	IssuerBaseURL       string
+	TokenEncryptionCode []byte
+	keys                KeyCache
+}
+
+func (kp *Service) Router() http.Handler {
+	mux := chi.NewRouter()
+	mux.Use(func(h http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			c, _ := httputil.DumpRequest(r, false)
+			log.WithField("req", string(c)).Debug("IDP request")
+			h.ServeHTTP(w, r)
+		})
+	})
+	mux.Handle(oidc.DiscoveryEndpoint, http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
```

Comment:
I think if you extract the handler into a function, you can more easily test the handler in your test cases

Handle will handle all HTTP Methods, is that really what we want here? It seems like we want only Post, no?